### ISSUE TITLE
BUG: Major performance issue (45x) with np.convolve on win 11 vs linux or MKL numpy

### ISSUE THREAD

**Issue by Niwotian60** [numpy/numpy#22821]
### Describe the issue:

Running np.convolve with the default pip numpy install (libopenblas) runs a simple test 45x slower on Win11 than the identical  code on the same machine in a WSL2 Debian instance or on the same machine with a different installation of numpy compiled with the intel MKL libraries. I first encountered this in numpy 1.23.4 (it probably existed prior to this) and still manifests itself in 1.24.0.

The timing for the simple code (below) is ~25s/loop on Win 11 vs 546ms/loop on linux (same machine running as a WSL2 debian virtual machine) and 433ms/loop on Win 11 with the MKL library build.

Importing ThreadpoolController reports the following:
```
In [3]: threadpool_info() 
Out[3]: [{'user_api': 'blas', 'internal_api': 'openblas', 'prefix': 'libopenblas', 'filepath': 'C:\python\Lib\site-packages\numpy\.libs\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll', 'version': '0.3.20', 'threading_layer': 'pthreads', 'architecture': 'Haswell', 'num_threads': 20}]
In [4]: tc = ThreadpoolController() 
In [5]: a1,a2=np.random.random(100000), np.random.random(100000)
In [6]: %timeit np.convolve(a1,a2) 
25.2 s ± 160 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) 
In [7] tc.info() 
Out[7]: [{'user_api': 'blas', 'internal_api': 'openblas', 'prefix': 'libopenblas', 'filepath': 'C:\python\Lib\site-packages\numpy\.libs\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll', 'version': '0.3.20', 'threading_layer': 'pthreads', 'architecture': 'Haswell', 'num_threads': 20}]

The reported num_threads matches my cpu (10/20 core). Watching the task manager performance monitor during the program run showed a very strong affinity to a single CPU core (~100%) while the others were at noise level (0-10% utilization).

Just as a reference point, running the same code with the same install but using `scipy.signal.fftconvolve` gives:
`8.54 ms ± 42.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`
``````
System Info:
```
Processor	Intel(R) Core(TM) i9-10900KF CPU @ 3.70GHz   3.70 GHz
Installed RAM	64.0 GB (61.8 GB usable)
GPU                 Nvidia GeForce RTX 3090 (24GB VRAM)
```

``

### Reproduce the code example:

```python
>>> import numpy as np
>>> sample_size = 100_000
>>> a1, a2 = np.random.random(sample_size), np.random.random(sample_size)
>>> %timeit np.convolve(a1,a2)
25.1 s ± 76.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### Error message:

```shell
No errors, just major (45x) performance issue
```

### NumPy/Python version information:

Python 3.11.1 installed from Python home page
Numpy 1.24.0 (and 1.23.5 and 1.23.4, probably earlier versions as well)

### Context for the issue:

No direct impact to me due to multiple alternatives. Not sure how pervasive the issue is, though.

**Comment by mattip** [comment]
> Watching the task manager performance monitor during the program run showed a very strong affinity to a single CPU core (~100%) while the others were at noise level (0-10% utilization).

This sounds like something is off with the threading code. Do you by any chance have environment variables set that would limit OpenBLAS to one thread, like `OMP_NUM_THREADS` ? Anything suspicious in the output of `> SET`?

**Comment by Niwotian60** [comment]
II don't see any environment variables that should affect numpy. AFAIK from what I can see the only environment I have set that impacts python is PYTHONPATH and CONDA_PROMPT_MODIFIER (which I believe vscode set since I don't run conda).

I'm not entirely sure what you mean by "> SET" (is this a reference to the old console set command in Windows (I use PowerShell)? If so, the answer is my response above.

**Comment by mattip** [comment]
The easiest explanation for the slowdown is that something is restricting OpenBLAS from using more threads. I wonder what it could be.

**Comment by Niwotian60** [comment]
I agree, although I think it may be more than that. given the time delta. Even if all 20 threads were fully utilized (and bearing in mind that 10 of the CPUs are hyperthreads), I wouldn't expect the time to decrease by a factor of 45 (superlinear). One other thing I should mention is that in trying to understand the issue myself and rule out any environmental or configuration issues I 
uninstalled numpy and everything related to it and then reinstalled using the semi-deprecated --no-binary flag. Executing np.__config__.show() indicated no BLAS libraries (as expected)) and while the timings were bad, they were still much better than the default BLAS install (~8.4s/loop vs 25s).

Btw, thanks for the replies and help. It is all too easy in this medium to forget that there is an underlying social interaction and to fail to show my appreciation!

**Comment by seberg** [comment]
What does `threadpoolctl` show in debian for the openblas install?  Maybe there is some more clear difference visible.  I am slightly suprised you would see "Haswell" on your CPU, but don't have a computer for comparison to see what is expected.

**Comment by Niwotian60** [comment]
I ran the following script in ipython in debian and win11 (same machine, WLS2 debian instance) with seemingly identical results:

```
from pprint import pprint
from threadpoolctl import threadpool_info
pprint(threadpool_info())
import numpy as np
pprint(threadpool_info())
```
Results for debian:
```
[]
[{'architecture': 'Haswell',
  'filepath': '/home/linuxbrew/.linuxbrew/lib/python3.10/site-packages/numpy.libs/libopenblas64_p-r0-15028c96.3.21.so',
  'internal_api': 'openblas',
  'num_threads': 20,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.21'}]
```
Results for Win11:
```
[]
[{'architecture': 'Haswell',
  'filepath': 'C:\\python\\Lib\\site-packages\\numpy\\.libs\\libopenblas64__v0.3.21-gcc_10_3_0.dll',
  'internal_api': 'openblas',
  'num_threads': 20,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.21'}]
```

**Comment by tylerjereddy** [comment]
The cross-listed SciPy ticket may be related. I looks like Evgeni and I could repro to a lesser extent on Linux.

I had success controlling the threads used by the blas backend to completely eliminate the problem in a minimum viable reproducer for the performance gap.
